common/reactive/src/main/java/io/helidon/common/reactive/MultiConcatArray.java

```diff
@@ -35,31 +36,89 @@
     public void subscribe(Flow.Subscriber<? super T> subscriber) {
         ConcatArraySubscriber<T> parent = new ConcatArraySubscriber<>(subscriber, sources);
         subscriber.onSubscribe(parent);
-        parent.nextSource();
+        parent.nextSource(parent.produced);
     }
 
-    static final class ConcatArraySubscriber<T> extends SubscriptionArbiter
-    implements Flow.Subscriber<T> {
+    protected static final class ConcatArraySubscriber<T>
+    implements Flow.Subscriber<T>, Flow.Subscription {
 
         private final Flow.Subscriber<? super T> downstream;
 
         private final Flow.Publisher<T>[] sources;
 
-        private final AtomicInteger wip;
+        private Flow.Subscription subscription;
 
         private int index;
 
-        private long produced;
+        private long produced = INIT;
+
+        private volatile long requested = SEE_OTHER;
+        private volatile long pending = INIT;
+        private volatile Thread lastThreadCompleting;
+        private boolean redo;
+
+        static final long BAD = Long.MIN_VALUE;
+        static final long CANCEL = Long.MIN_VALUE + 1;
+        static final long SEE_OTHER = Long.MIN_VALUE + 2;
+        static final long INIT = Long.MIN_VALUE + 3;
+
+        static final VarHandle REQUESTED;
+        static final VarHandle PENDING;
+        static final VarHandle LASTTHREADCOMPLETING;
+
+        static {
+            try {
+                MethodHandles.Lookup lookup = MethodHandles.lookup();
+                REQUESTED = lookup.findVarHandle(ConcatArraySubscriber.class, "requested", long.class);
+                PENDING = lookup.findVarHandle(ConcatArraySubscriber.class, "pending", long.class);
+                LASTTHREADCOMPLETING = lookup
+                        .findVarHandle(ConcatArraySubscriber.class, "lastThreadCompleting", Thread.class);
+            } catch (Exception e) {
+                throw new Error("Expected lookup to succeed", e);
+            }
+        }
 
         ConcatArraySubscriber(Flow.Subscriber<? super T> downstream, Flow.Publisher<T>[] sources) {
             this.downstream = downstream;
             this.sources = sources;
-            this.wip = new AtomicInteger();
         }
 
         @Override
         public void onSubscribe(Flow.Subscription subscription) {
-            super.setSubscription(subscription);
+            long p0 = pending;
+            if (p0 == CANCEL) {
+               subscription.cancel();
+               return;
+            }
+
+            produced++; // assert: matching request(1) has been done by nextSource()
+            this.subscription = subscription;
+            // assert: requested == SEE_OTHER
+            REQUESTED.setOpaque(this, p0); // assert: p0 is guaranteed to be a value of requested never seen before
```

Comment:
Two things need doing:
`long oldProduced = produced;` before this line, then use `oldProduced` to the end of this method.

This addresses the case raised by @akarnok - concurrent `request` used to update `produced`.

and

`REQUESTED.setVolatile(this, p0)` to replace `setOpaque` - opaque doesn't guarantee ordering with other stores.